### PR TITLE
osd/PrimaryLogPG: Refactor do_osd_ops

### PR DESCRIPTION
**Pull request by markhpc**
This is a first pass refactoring do_osd_ops to move the contents of the giant for loop / switch statement into separate sub functions.  Selfishly the goal here is to make wallclock profiling of ops more useful, but also to make the code easier to read and more maintainable as well.

This PR now is passing local performance tests and is ready for QA.

Signed-off-by: Mark Nelson <mnelson@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/osd/PrimaryLogPG.h [hunk 1]
```diff
@@ -1504,8 +1504,74 @@ class PrimaryLogPG : public PG, public PGBackend::Listener {
   void snap_trimmer(epoch_t e) override;
   void kick_snap_trim() override;
   void snap_trimmer_scrub_complete() override;
+
+  bool _do_osd_op_pre(OpContext *ctx, OSDOp& osd_op, int& result);
```
Comment (jdurgin) [inline src/osd/PrimaryLogPG.h]: nit: the result args for _do_osd_op_{pre,post} would be clearer if they were pointers rather than references

**Comment by jdurgin** [comment]
looks like a good way to split things to me - simpler than I thought with everything in OpContext and OSDOp

**Comment by markhpc** [comment]
Example output during RGW bucket listing (showing that all time in do_osd_ops is actually spent in CEPH_OSD_OP_OMAPGETVALS):

```
          |           |                 + 60.00% rgw_bucket_list
          |           |                   + 29.80% get_obj_vals
          |           |                   | + 29.80% cls_cxx_map_get_vals
          |           |                   |   + 25.60% PrimaryLogPG::do_osd_ops
          |           |                   |   | + 25.60% PrimaryLogPG::_ceph_osd_op_omapgetvals
          |           |                   |   |   + 13.00% BlueStore::OmapIteratorImpl::next
          |           |                   |   |   | + 9.60% next
          |           |                   |   |   | | + 9.60% RocksDBStore::RocksDBWholeSpaceIteratorImpl::next
          |           |                   |   |   | |   + 9.40% rocksdb::DBIter::Next
          |           |                   |   |   | |     + 9.00% rocksdb::MergingIterator::Next
          |           |                   |   |   | |     | + 8.60% Next
          |           |                   |   |   | |     | | + 8.60% rocksdb::(anonymous namespace)::LevelIterator::Next
          |           |                   |   |   | |     | |   + 8.60% Next
          |           |                   |   |   | |     | |     + 8.00% rocksdb::BlockBasedTableIterator<rocksdb::DataBlockIter, rocksdb::Slice>::FindKeyForward
          |           |                   |   |   | |     | |     | + 7.80% rocksdb::BlockBasedTableIterator<rocksdb::DataBlockIter, rocksdb::Slice>::InitDataBlock
          |           |                   |   |   | |     | |     | | + 6.60% Prefetch
          |           |                   |   |   | |     | |     | | | + 6.60% BlueRocksRandomAccessFile::Prefetch
          |           |                   |   |   | |     | |     | | |   + 6.60% read
          |           |                   |   |   | |     | |     | | |     + 6.60% BlueFS::_read
          |           |                   |   |   | |     | |     | | |       + 3.40% lock
          |           |                   |   |   | |     | |     | | |       + 2.00% KernelDevice::read
          |           |                   |   |   | |     | |     | | |       + 0.80% shared_lock
          |           |                   |   |   | |     | |     | | |       + 0.20% unlock
          |           |                   |   |   | |     | |     | | |       + 0.20% clear
          |           |                   |   |   | |     | |     | | + 1.20% rocksdb::BlockBasedTable::NewDataBlockIterator<rocksdb::DataBlockIter>
          |           |                   |   |   | |     | |     | + 0.20% ResetDataIter
          |           |                   |   |   | |     | |     + 0.60% rocksdb::BlockBasedTableIterator<rocksdb::DataBlockIter, rocksdb::Slice>::Next
          |           |                   |   |   | |     | + 0.20% pop
          |           |                   |   |   | |     + 0.40% FindNextUserEntry
          |           |                   |   |   | + 2.00% BlueStore::log_latency
          |           |                   |   |   | + 0.60% ~shared_lock
          |           |                   |   |   | + 0.40% shared_lock
          |           |                   |   |   | + 0.40% now
          |           |                   |   |   + 5.40% BlueStore::OmapIteratorImpl::value
          |           |                   |   |   | + 4.60% value
          |           |                   |   |   | + 0.60% ~shared_lock
          |           |                   |   |   | + 0.20% shared_lock
          |           |                   |   |   + 3.80% BlueStore::OmapIteratorImpl::key
          |           |                   |   |   + 1.00% BlueStore::OmapIteratorImpl::valid
          |           |                   |   |   + 0.80% encode
          |           |                   |   |   + 0.60% ~basic_string
          |           |                   |   |   + 0.40% ceph::buffer::v14_2_0::list::claim_append
          |           |                   |   |   + 0.20% ~list
          |           |                   |   |   + 0.20% BlueStore::get_omap_iterator
          |           |                   |   + 2.80% ceph::decode<std::map<std::string, ceph::buffer::v14_2_0::list, std::less<std::string>, std::allocator<std::pair<std::string const, ceph::buffer::v14_2_0::list> > >, denc_traits<std::map<std::string, ceph::buffer::v14_2_0::list, std::less<std::string>, std::allocator<std::pair<std::string const, ceph::buffer::v14_2_0::list> > >, void> >
          |           |                   |   + 1.20% ~vector
          |           |                   |   + 0.20% encode
          |           |                   + 8.40% cls_log
          |           |                   | + 7.80% vsnprintf
          |           |                   | | + 7.60% __vsnprintf_chk
          |           |                   | |   + 7.20% vfprintf
          |           |                   | |   | + 0.60% _itoa_word
          |           |                   | |   | + 0.40% free@plt
          |           |                   | |   | + 0.40% __GI__IO_default_xsputn
          |           |                   | |   | + 0.20% tc_free
          |           |                   | |   | + 0.20% strchrnul
          |           |                   | |   + 0.20% _IO_no_init
          |           |                   | + 0.20% operator()<CephContext*>
          |           |                   + 6.80% decode
          |           |                   | + 6.80% rgw_bucket_dir_entry::decode
          |           |                   |   + 5.20% decode
          |           |                   |   | + 4.00% rgw_bucket_dir_entry_meta::decode
          |           |                   |   | + 0.40% decode_raw<unsigned char>
          |           |                   |   | + 0.20% rgw_bucket_entry_ver::decode
          |           |                   |   | + 0.20% decode_raw<ceph_le<unsigned short> >
          |           |                   |   | + 0.20% decode_raw<ceph_le<unsigned long long> >
          |           |                   |   | + 0.20% decode_raw<ceph_le<unsigned int> >
          |           |                   |   + 1.20% ceph::decode
          |           |                   |   + 0.40% decode_packed_val<unsigned long>
          |           |                   + 4.80% operator[]
          |           |                   + 4.00% encode
          |           |                   + 1.80% ~rgw_cls_list_ret
          |           |                   + 1.00% ~map
          |           |                   + 1.00% operator=
          |           |                   + 0.60% rgw_bucket_dir_entry::~rgw_bucket_dir_entry
          |           |                   + 0.60% cbegin
          |           |                   + 0.20% rgw_bucket_dir_entry
          |           |                   + 0.20% operator==<char>
          |           |                   + 0.20% operator++
          |           |                   + 0.20% decode_list_index_key
```

**Comment by markhpc** [comment]
First Teuthology run uncovered a bug in the refactor that lead to segmentation faults in the OSD:

http://pulpito.ceph.com/nhm-2019-09-19_21:47:31-rados-wip-mark-testing-2019-09-19-1056-distro-basic-smithi/

However after this was fixed, a subsequent run of the failed jobs fixed all OSD related segfaults but still shows a single abort failure in bluestore _fsck and segmentation faults in ceph_test_rados_api_watch notify_pp (need to investigate if either of these are related?)

http://pulpito.ceph.com/nhm-2019-09-22_21:45:39-rados-wip-mark-testing-2019-09-20-1503-distro-basic-smithi/

**Comment by markhpc** [comment]
Still intend to get back to this, but perhaps break this up until smaller PRs.

**Comment by markhpc** [comment]
Not dead yet, but probably should be broken up.

**Pull request closed without merging**